src/ctia/task/migration/store.clj

```diff
@@ -256,10 +259,135 @@ Rollover requires refresh so we cannot just call ES with condition since refresh
    batch-size :- s/Int
    migrated-count :- s/Int]
-  (when rollover?
+  (when (rollover? aliased max_docs batch-size migrated-count)
     (es-index/refresh! conn write-index)
     (rollover-store (store-map->es-conn-state store-map))))
 
+(s/defn missing-query :- ESQuery
+  "implement missing filter through a must_not exists bool query
+  https://www.elastic.co/guide/en/elasticsearch/reference/5.6/query-dsl-exists-query.html#missing-query"
+  [field]
+  {:bool
+   {:must_not
+    {:exists
+     {:field field}}}})
+
+(s/defn range-query :- ESQuery
+  "returns a bool range filter query with start and end limits"
+  [date field unit]
+  {:bool
+   {:filter
+    {:range
+     {field
+      {:gte date
+       :lt (str date "||+1" unit)}}}}})
+
+(s/defn last-range-query :- ESQuery
+  "returns a bool range filter query with only start limit. Add epoch_millis
+  format if specified so (required  when date comes from search_after)"
+  ([date field epoch-millis?]
+   (last-range-query date field epoch-millis? false))
+  ([date field epoch-millis? strict?]
+   {:bool
+    {:filter
+     {:range
+      {field
+       (cond-> {:gte date}
+         strict? (clojure.set/rename-keys {:gte :gt})
+         epoch-millis? (assoc :format "epoch_millis"))}}}}))
+
+(def missing-date-str "2010-01-01T00:00:00.000Z")
+(def missing-date (time-coerce/to-date-time missing-date-str))
+
+(defn missing-bucket?
+  "Returns true if the date is related to missing date.
+   In aggregation query, we assign a default date, `missing-date` as default value
+   for documents that do not have the datetime field on which is based the aggregation.
+   Depending on date interval the date of missing bucket returned by ES could be
+   either equal or before missing date. Thus we test this in order to not only rely on
+   the value returned by ES aggregation and prevent modification on the aggregation
+   between ES versions."
+  [date-str]
+  (or (= date-str missing-date-str)
+      (-> (time-coerce/to-date-time date-str)
+          (time/before? missing-date))))
+
+(def Interval (s/enum "year" "month" "week" "day"))
+
+(s/defn format-buckets :- (s/maybe [ESQuery])
+  "format buckets from aggregation results into an ordered list of proper bool queries"
+  [raw-buckets :- [(st/open-schema
+                    {:doc_count s/Int
+                     :key_as_string s/Str})]
+   field :- s/Keyword
+   interval :- Interval]
+  (let [unit (case interval
+               "year" "y"
+               "month" "M"
+               "week" "w"
+               "day" "d")
+        filtered (->> raw-buckets
+                      (filter #(< 0 (:doc_count %)))
+                      (map :key_as_string))
+        [first-date & rest-dates] filtered
+        queries (map #(range-query % field unit)
+                     (drop-last filtered))
+        last-query (some-> (last filtered)
+                           (last-range-query field false))]
+    (cond-> queries
+      last-query (concat [last-query])
```

Comment:
Why do you use `concat` instead of `(conj last-query)`?

Reply:
queries is a `list`.
``` clojure
;; notice conjoining to a list is done at the beginning
(conj '(1 2 3) 4)
;;=> (4 1 2 3)
```
I could use `mapv` for building queries to then use `conj`.